internal/platform/merkletree/sparsemt/tree.go

```diff
@@ -0,0 +1,157 @@
+package sparsemt
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/freeverseio/laos-universal-node/internal/platform/merkletree"
+	"github.com/freeverseio/laos-universal-node/internal/platform/storage"
+)
+
+// Tree implements sparse merkle tree
+type Tree struct {
+	store   storage.Tx
+	prefix  string
+	depth   uint
+	root    common.Hash
+	nLeaves *big.Int
+}
+
+// New instantiates a new merkle tree with storage capability, depth and prefix
+func New(depth uint, store storage.Tx, prefix string) (merkletree.MerkleTree, error) {
+	if store == nil {
+		return nil, errors.New("store is nil")
+	}
+	return &Tree{
+		store:   store,
+		depth:   depth,
+		nLeaves: new(big.Int).Exp(big.NewInt(2), big.NewInt(int64(depth)), nil),
+		prefix:  prefix,
+	}, nil
+}
+
+// SetRoot sets the root of the tree to "hash"
+func (b *Tree) SetRoot(hash common.Hash) {
+	b.root = hash
+}
+
+// Root returns the root of the tree
+func (b *Tree) Root() common.Hash {
+	return b.root
+}
+
+// Proof calculates the proofs from leaf at index "idx" going up to the root
+func (b *Tree) Proof(idx *big.Int) ([]common.Hash, error) {
```

Comment:
maybe add some comments for documentation?

Reply:
added